File: src/main/java/net/slipp/social/security/Sha256ToBCryptPasswordEncoder.java
package net.slipp.social.security;

import org.springframework.security.authentication.encoding.MessageDigestPasswordEncoder;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;

public class Sha256ToBCryptPasswordEncoder implements PasswordEncoder {
    private PasswordEncoder bcryptPasswordEncoder;
    private MessageDigestPasswordEncoder sha256PasswordEncoder;

    @Override
    public String encode(CharSequence rawPassword) {
        return bcryptPasswordEncoder.encode(rawPassword);
    }

    @Override
    public boolean matches(CharSequence rawPassword, String encodedPassword) {
        if (encodedPassword.startsWith("$2a$10$") && encodedPassword.length() == 60) {
            return bcryptPasswordEncoder.matches(rawPassword, encodedPassword);
        }

        if (encodedPassword.length() == 64) {
            return sha256PasswordEncoder.isPasswordValid(encodedPassword, rawPassword.toString(), "");
        }

        return false;
    }

    public void setBcryptPasswordEncoder(PasswordEncoder bcryptPasswordEncoder) {
        this.bcryptPasswordEncoder = bcryptPasswordEncoder;
    }

    public void setSha256PasswordEncoder(MessageDigestPasswordEncoder sha256PasswordEncoder) {
        this.sha256PasswordEncoder = sha256PasswordEncoder;
    }
}

File: src/main/scala/net/slipp/social/security/Sha256ToBCryptPasswordEncoder.scala
package net.slipp.social.security

import org.springframework.security.authentication.encoding.MessageDigestPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder

class Sha256ToBCryptPasswordEncoder extends PasswordEncoder {
  private var bcryptPasswordEncoder: PasswordEncoder = null
  private var sha256PasswordEncoder: MessageDigestPasswordEncoder = null

  def encode(rawPassword: CharSequence): String = {
    return bcryptPasswordEncoder.encode(rawPassword)
  }

  def matches(rawPassword: CharSequence, encodedPassword: String): Boolean = {
    if (encodedPassword.startsWith("$2a$10$") && encodedPassword.length == 60) {
      return bcryptPasswordEncoder.matches(rawPassword, encodedPassword)
    }
    if (encodedPassword.length == 64) {
      return sha256PasswordEncoder.isPasswordValid(encodedPassword, rawPassword.toString, null)
    }
    return false
  }

  def setBcryptPasswordEncoder(bcryptPasswordEncoder: PasswordEncoder) {
    this.bcryptPasswordEncoder = bcryptPasswordEncoder
  }

  def setSha256PasswordEncoder(sha256PasswordEncoder: MessageDigestPasswordEncoder) {
    this.sha256PasswordEncoder = sha256PasswordEncoder
  }
}
